pkg/blobinfocache/internal/prioritize/prioritize.go

```diff
@@ -77,34 +82,64 @@ func (css *candidateSortState) Swap(i, j int) {
 	css.cs[i], css.cs[j] = css.cs[j], css.cs[i]
 }
 
-// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with a parameter for the
-// number of entries to limit, only to make testing simpler.
-func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, maxCandidates int) []blobinfocache.BICReplacementCandidate2 {
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// destructivelyPrioritizeReplacementCandidatesWithMax is destructivelyPrioritizeReplacementCandidates with parameters for the
+// number of entries to limit for known and unknown location separately, only to make testing simpler.
+// TODO: following function is not destructive any more in the nature instead priortized result is actually copies of the original
+// candidate set, so In future we might wanna re-name this public API and remove the destructive prefix.
+func destructivelyPrioritizeReplacementCandidatesWithMax(cs []CandidateWithTime, primaryDigest, uncompressedDigest digest.Digest, totalLimit int, noLocationLimit int) []blobinfocache.BICReplacementCandidate2 {
+	// split unknown candidates and known candidates
+	// and limit them seperately.
+	var knownLocationCandidates []CandidateWithTime
+	var unknownLocationCandidates []CandidateWithTime
+	for _, candidate := range cs {
+		if candidate.Candidate.UnknownLocation {
+			unknownLocationCandidates = append(unknownLocationCandidates, candidate)
+		} else {
+			knownLocationCandidates = append(knownLocationCandidates, candidate)
+		}
+	}
+
 	// We don't need to use sort.Stable() because nanosecond timestamps are (presumably?) unique, so no two elements should
 	// compare equal.
 	sort.Sort(&candidateSortState{
-		cs:                 cs,
+		cs:                 knownLocationCandidates,
 		primaryDigest:      primaryDigest,
 		uncompressedDigest: uncompressedDigest,
 	})
 
-	resLength := len(cs)
-	if resLength > maxCandidates {
-		resLength = maxCandidates
+	knownLocationCandidatesUsed := min(len(knownLocationCandidates), totalLimit)
+	remainingCapacity := totalLimit - knownLocationCandidatesUsed
+	unknownLocationCandidatesUsed := 0
+	if len(unknownLocationCandidates) > 0 {
+		unknownLocationCandidatesUsed = min(noLocationLimit, min(remainingCapacity, len(unknownLocationCandidates)))
 	}
-	res := make([]blobinfocache.BICReplacementCandidate2, resLength)
-	for i := range res {
-		res[i] = cs[i].Candidate
+	res := make([]blobinfocache.BICReplacementCandidate2, knownLocationCandidatesUsed)
+	for i := 0; i < knownLocationCandidatesUsed; i++ {
+		res[i] = knownLocationCandidates[i].Candidate
+	}
+	// If candidates with unknown location are found, lets add them to final list
+	if len(unknownLocationCandidates) > 0 {
```

Comment:
This condition seems unnecessary; the loop below handles `unknownLocationCandidatesUsed == 0` already.